crates/rate_limit/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+pub async fn token_bucket(
+    mem_cache: &Cache<String, u64>,
+    redis_write_pool: &Arc<Pool>,
+    key: String,
+    max_tokens: u32,
+    interval: Duration,
+    refill_rate: u32,
+) -> Result<(), RateLimitError> {
+    // Check if the key is in the memory cache of rate limited keys
+    // to omit the redis RTT in case of flood
+    if let Some(reset) = mem_cache.get(&key).await {
+        return Err(RateLimitError::RateLimitExceeded(RateLimitExceeded {
+            reset,
+        }));
+    }
+
+    let result = token_bucket_many(
+        redis_write_pool,
+        vec![key.clone()],
+        max_tokens,
+        interval,
+        refill_rate,
+    )
+    .await
+    .map_err(RateLimitError::Internal)?;
+
+    let (remaining, reset) = result.get(&key).expect("Should contain the key");
+    if remaining.is_negative() {
+        let reset_interval = reset / 1000;
+
+        // Insert the rate-limited key into the memory cache to avoid the redis RTT in
+        // case of flood
+        mem_cache.insert(key, reset_interval).await;
+
+        Err(RateLimitError::RateLimitExceeded(RateLimitExceeded {
+            reset: reset_interval,
+        }))
+    } else {
+        Ok(())
+    }
+}
+
+/// Rate limit check using a token bucket algorithm for many keys.
+pub async fn token_bucket_many(
+    redis_write_pool: &Arc<Pool>,
+    keys: Vec<String>,
+    max_tokens: u32,
+    interval: Duration,
+    refill_rate: u32,
+) -> Result<HashMap<String, (i64, u64)>, InternalRateLimitError> {
+    let now = Utc::now();
+
+    // Remaining is number of tokens remaining. -1 for rate limited.
+    // Reset is the time at which there will be 1 more token than before. This
+    // could, for example, be used to cache a 0 token count.
+    Script::new(include_str!("token_bucket.lua"))
+        .key(keys)
+        .arg(max_tokens)
+        .arg(interval.num_milliseconds())
+        .arg(refill_rate)
+        .arg(now.timestamp_millis())
+        .invoke_async::<_, String>(
+            &mut redis_write_pool
+                .clone()
+                .get()
+                .await
+                .map_err(InternalRateLimitError::Pool)?,
+        )
+        .await
+        .map_err(InternalRateLimitError::Redis)
+        .map(|value| serde_json::from_str(&value).expect("Redis script should return valid JSON"))
+}
+
+#[cfg(test)]
+mod tests {
+    const REDIS_URI: &str = "redis://localhost:6379";
+    const REFILL_INTERVAL_MILLIS: i64 = 100;
+
+    use {
+        super::*,
+        deadpool_redis::{Config, Runtime},
+        redis::AsyncCommands,
+        tokio::time::sleep,
+    };
+
+    async fn redis_clear_keys(conn_uri: &str, keys: &[String]) {
+        let client = redis::Client::open(conn_uri).unwrap();
+        let mut conn = client.get_async_connection().await.unwrap();
+        for key in keys {
+            let _: () = conn.del(key).await.unwrap();
+        }
+    }
+
+    #[tokio::test]
+    async fn test_token_bucket_many() {
```

Comment:
This test has significantly less coverage than [the one from Notify Server](https://github.com/WalletConnect/notify-server/blob/660a71ae88911965ce7657e7bb025ecd37f9e028/tests/integration.rs#L1813). Can we copy those cases directly so we can keep the level of coverage?

Missing:
- refill_in calculation
- checking that tokens recover 1 at a time rather than all at once
- test for [separate keys](https://github.com/WalletConnect/notify-server/blob/660a71ae88911965ce7657e7bb025ecd37f9e028/tests/integration.rs#L1878)

Why did you remove the ability to mock the clock?

Reply:
Makes sense to update the test coverage with the missing ones, thanks for pointing out this!

> Why did you remove the ability to mock the clock? That helps ensure that tests aren't flaky as well as I think making the refill_in test case possible.

Maybe it's better to pass the timestamp to the function instead of hacking the clock for the test (it looks a bit hacky to me). We can just pass `Utc::now().timestamp_millis()` to the function instead of getting it inside of it.

I have updated tests and the function in the https://github.com/WalletConnect/utils-rs/pull/14/commits/729ec960a436fb748cd70f9fc9a24f3259d75cd7. 

The test coverage now has:
*  [refill_in calculation](https://github.com/WalletConnect/utils-rs/pull/14/files#diff-4bb737c17f7e69d6f041e206ed5a2e6ea8c936e0eb108088d3acb80956c001bbR175) assertions,
* [checking that tokens recover 1 at a time](https://github.com/WalletConnect/utils-rs/pull/14/files#diff-4bb737c17f7e69d6f041e206ed5a2e6ea8c936e0eb108088d3acb80956c001bbR193-R199),
* [test for multiple keys](https://github.com/WalletConnect/utils-rs/pull/14/files#diff-4bb737c17f7e69d6f041e206ed5a2e6ea8c936e0eb108088d3acb80956c001bbR203-R217) running in parallel.

Reply:
Are you planning to replace the Notify Server implementation with this standard one? If so, the mocked clock is was necessary to be able to test that rate limits are applied to regular endpoints.

It's common practice to mock clocks, I don't see what's "hacky" about that.

Reply:
I want to apply this unified implementation to the Blockchain-api and Echo server and then probably to the Notify (not on the table yet).